lib/fluent/event.rb

```diff
@@ -267,6 +274,53 @@ def to_msgpack_stream(time_int: false)
     end
   end
 
+  class CompressedMessagePackEventStream < MessagePackEventStream
+    def initialize(data, cached_unpacker = nil, size = 0, unpacked_times: nil, unpacked_records: nil)
+      super
+      @decompressed_data = nil
+      @compressed_data = data
+    end
+
+    def dup
```

Comment:
Ah... It's better to fix original implementation `MessagePackEventStream#dup` to use `self.class` instead of re-implementing this method in subclass.